File: src/web/Controllers/RoutinesController.cs
using System;
using System.Threading.Tasks;
using core.fs.Routines;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using web.Utils;

namespace web.Controllers;

[ApiController]
[Authorize]
[Route("api/[controller]")]
public class RoutinesController(Handler handler) : ControllerBase
{
    [HttpGet]
    public Task<ActionResult> GetRoutines() =>
        this.OkOrError(handler.Handle(new Query(User.Identifier())));

    [HttpPost]
    public Task<ActionResult> CreateRoutine([FromBody] Create command) =>
        this.OkOrError(
            handler.HandleCreate(
                User.Identifier(),
                command
            )
        );

    [HttpPut("{id}")]
    public Task UpdateRoutine([FromBody]Update command) =>
        this.OkOrError(
            handler.HandleUpdate(
                User.Identifier(),
                command
            )
        );

    [HttpDelete("{id}")]
    public Task DeleteRoutine([FromRoute]Guid id) =>
        this.OkOrError(
            handler.Handle(
                new Delete(User.Identifier(), id)
            )
        );

    [HttpPut("{id}/steps")]
    public Task<ActionResult> AddRoutineStep([FromBody]AddStep command) =>
        this.OkOrError(
            handler.HandleAddStep(
                User.Identifier(), command
            )
        );
    
    [HttpPost("{id}/steps/{stepIndex}")]
    public Task<ActionResult> UpdateRoutineStep([FromBody]UpdateStep command) =>
        this.OkOrError(
            handler.HandleUpdateStep(
                User.Identifier(), command
            )
        );


    [HttpDelete("{id}/steps/{stepIndex}")]
    public Task<ActionResult> RemoveRoutineStep([FromRoute] Guid id, [FromRoute] int stepIndex) =>
        this.OkOrError(
            handler.Handle(
                new RemoveStep(id, stepIndex, User.Identifier())
            )
        );

    [HttpPost("{id}/steps/{stepIndex}/position")]
    public Task<ActionResult> MoveRoutineStep([FromBody] MoveStep cmd) =>
        this.OkOrError(
            handler.HandleMoveStep(
                User.Identifier(), cmd
            )
        );
}

File: src/web.fs/Controllers/RoutinesController.fs
namespace web.Controllers

open System
open System.Threading.Tasks
open Microsoft.AspNetCore.Mvc
open Microsoft.AspNetCore.Authorization
open core.fs.Routines
open web.Utils

[<ApiController>]
[<Authorize>]
[<Route("api/[controller]")>]
type RoutinesController(handler: Handler) =
    inherit ControllerBase()

    [<HttpGet>]
    member this.GetRoutines() =
        this.OkOrError(handler.Handle({Query.UserId = this.User.Identifier()}))

    [<HttpPost>]
    member this.CreateRoutine([<FromBody>] command: Create) =
        this.OkOrError(handler.HandleCreate (this.User.Identifier()) command)

    [<HttpPut("{id}")>]
    member this.UpdateRoutine([<FromBody>] command: Update) =
        this.OkOrError(handler.HandleUpdate (this.User.Identifier()) command)

    [<HttpDelete("{id}")>]
    member this.DeleteRoutine([<FromRoute>] id: Guid) =
        this.OkOrError(handler.Handle({Delete.UserId = this.User.Identifier(); Id = id}))

    [<HttpPut("{id}/steps")>]
    member this.AddRoutineStep([<FromBody>] command: AddStep) =
        this.OkOrError(handler.HandleAddStep (this.User.Identifier()) command)

    [<HttpPost("{id}/steps/{stepIndex}")>]
    member this.UpdateRoutineStep([<FromBody>] command: UpdateStep) =
        this.OkOrError(handler.HandleUpdateStep (this.User.Identifier()) command)

    [<HttpDelete("{id}/steps/{stepIndex}")>]
    member this.RemoveRoutineStep([<FromRoute>] id: Guid, [<FromRoute>] stepIndex: int) =
        this.OkOrError(handler.Handle({RemoveStep.Id = id; StepIndex = Nullable(stepIndex); UserId = this.User.Identifier()}))

    [<HttpPost("{id}/steps/{stepIndex}/position")>]
    member this.MoveRoutineStep([<FromBody>] cmd: MoveStep) =
        this.OkOrError(handler.HandleMoveStep (this.User.Identifier()) cmd)
